File: EX2_Flavio/TransitSoftPersistance/src/main/java/pe/edu/pucp/softinv/daoImp/VehiculoDAOImpl.java
package pe.edu.pucp.softinv.daoImp;

import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;
import pe.pucp.transitsoft.dao.VehiculoDAO;
import pe.pucp.transitsoft.daoImp.util.Columna;
import pe.pucp.transitsoft.model.PropietarioDTO;
import pe.pucp.transitsoft.model.VehiculoDTO;
import pe.pucp.transitsoft.util.MotorDeBaseDeDatos;

public class VehiculoDAOImpl extends DAOImplBase implements VehiculoDAO {

    private VehiculoDTO vehiculoDTO;
    private int conPropietario;
            
    public VehiculoDAOImpl() {
        super("vehiculo");
        this.vehiculoDTO = null;
        this.retornarLlavePrimaria = true;
        this.conPropietario = 0;
    }

    @Override
    protected void configurarListaDeColumnas() {
        this.listaColumnas.add(new Columna("id", true, true));
        this.listaColumnas.add(new Columna("placa", false, false));
        this.listaColumnas.add(new Columna("marca", false, false));
        this.listaColumnas.add(new Columna("modelo", false, false));
        this.listaColumnas.add(new Columna("anho", false, false));
    }

    @Override
    protected void instanciarObjetoDelResultSet() throws SQLException {
        this.vehiculoDTO = new VehiculoDTO();
        
        if (this.conPropietario == 1) {
            PropietarioDTO propietario = new PropietarioDTO();
            propietario.setId(this.resultSet.getInt("id_propietario"));
            propietario.setDni(this.resultSet.getString("dni"));
            propietario.setNombres(this.resultSet.getString("nombres"));
            propietario.setApellidos(this.resultSet.getString("apellidos"));
            propietario.setDireccion(this.resultSet.getString("direccion"));
            this.vehiculoDTO.setPropietario(propietario);
            
            this.vehiculoDTO.setId(this.resultSet.getInt("id_vehiculo"));
            this.vehiculoDTO.setPlaca(this.resultSet.getString("placa"));
            this.vehiculoDTO.setMarca(this.resultSet.getString("marca"));
            this.vehiculoDTO.setModelo(this.resultSet.getString("modelo"));
            this.vehiculoDTO.setAnho(this.resultSet.getInt("anho"));
            
            this.conPropietario = 0;
        } else {
            this.vehiculoDTO.setId(this.resultSet.getInt("id"));
            this.vehiculoDTO.setPlaca(this.resultSet.getString("placa"));
            this.vehiculoDTO.setMarca(this.resultSet.getString("marca"));
            this.vehiculoDTO.setModelo(this.resultSet.getString("modelo"));
            this.vehiculoDTO.setAnho(this.resultSet.getInt("anho"));
        }
    }

    @Override
    protected void incluirValorDeParametrosParaInsercion() throws SQLException {
        int i = 1;
        this.statement.setString(i++, vehiculoDTO.getPlaca());
        this.statement.setString(i++, vehiculoDTO.getMarca());
        this.statement.setString(i++, vehiculoDTO.getModelo());
        this.statement.setInt(i++, vehiculoDTO.getAnho());
    }
    
    @Override
    protected void incluirValorDeParametrosParaModificacion() throws SQLException  {
        int i = 1;
        this.statement.setString(i++, vehiculoDTO.getPlaca());
        this.statement.setString(i++, vehiculoDTO.getMarca());
        this.statement.setString(i++, vehiculoDTO.getModelo());
        this.statement.setInt(i++, vehiculoDTO.getAnho());
        this.statement.setInt(i++, vehiculoDTO.getId());
    }
    
    @Override
    protected void incluirValorDeParametrosParaEliminacion() throws SQLException  {
        int i = 1;
        this.statement.setInt(i++, vehiculoDTO.getId());
    }
    
    @Override
    protected void incluirValorDeParametrosParaObtenerPorId() throws SQLException  {
        int i = 1;
        this.statement.setInt(i++, vehiculoDTO.getId());
    }

    @Override
    protected void limpiarObjetoDelResultSet() {
        this.vehiculoDTO = null;
    }

    @Override
    protected void agregarObjetoALaLista(List lista) throws SQLException {
        this.instanciarObjetoDelResultSet();
        lista.add(this.vehiculoDTO);
    }

    @Override
    public ArrayList<VehiculoDTO> listarPorPropietario(String descripion_propietario) {
        String sql = this.generarSQLParaListarPorPropietario();
        this.conPropietario = 1;
        ArrayList<VehiculoDTO> listado = (ArrayList<VehiculoDTO>) super.listarTodos(sql, this::incluirValorDeParametrosParaListarPorPropietario, descripion_propietario);
        ArrayList<VehiculoDTO> listadoMSQL = (ArrayList<VehiculoDTO>) super.listarTodos(MotorDeBaseDeDatos.MSSQL, sql, this::incluirValorDeParametrosParaListarPorPropietario, descripion_propietario);
        if (listadoMSQL != null) {
            listado.addAll(listadoMSQL);
        }
        
        return listado;
    }

    private String generarSQLParaListarPorPropietario() {
        String sql = "SELECT v.id AS id_vehiculo, ";
        sql = sql.concat("v.placa, ");
        sql = sql.concat("v.marca, ");
        sql = sql.concat("v.modelo, ");
        sql = sql.concat("v.anho, ");
        sql = sql.concat("p.id AS id_propietario, ");
        sql = sql.concat("p.dni, ");
        sql = sql.concat("p.nombres, ");
        sql = sql.concat("p.apellidos, ");
        sql = sql.concat("p.direccion ");
        sql = sql.concat("FROM vehiculo v ");
        sql = sql.concat("JOIN vehiculo_propietario vp ON vp.id_vehiculo = v.id ");
        sql = sql.concat("JOIN propietario p ON p.id = vp.id_propietario ");
        sql = sql.concat("WHERE (p.dni LIKE ? OR p.nombres LIKE ? OR p.apellidos LIKE ?); ");
        return sql;
    }
    
    private void incluirValorDeParametrosParaListarPorPropietario(Object objetoParametros) {
        String cadena = (String) objetoParametros;
        if (cadena == null) cadena = "";
        cadena = cadena.trim();
        cadena = "%".concat(cadena).concat("%");
        try {
            this.statement.setString(1, cadena);
            this.statement.setString(2, cadena);
            this.statement.setString(3, cadena);
        } catch (SQLException ex) {
            Logger.getLogger(VehiculoDAOImpl.class.getName()).log(Level.SEVERE, null, ex);
        }
    }

    @Override
    public int insertar(VehiculoDTO vehiculo) {
        this.vehiculoDTO  = vehiculo;
        return super.insertar();
    }

    @Override
    public int modificar(VehiculoDTO vehiculo) {
        this.vehiculoDTO  = vehiculo;
        return super.modificar();
    }

    @Override
    public int eliminar(VehiculoDTO vehiculo) {
        this.vehiculoDTO  = vehiculo;
        return super.eliminar();
    }

    @Override
    public VehiculoDTO obtenerPorId(Integer id) {
        this.vehiculoDTO = new VehiculoDTO();
        this.vehiculoDTO.setId(id);
        super.obtenerPorId();
        return this.vehiculoDTO;
    }

    @Override
    public ArrayList<VehiculoDTO> listarTodos() {
        return (ArrayList<VehiculoDTO>)super.listarTodos();
    }
}

File: Antiguo/Lab09/TransitSoft/TransitSoftPersistance/daoImpl/VehiculoDAOImpl.cs
using System;
using System.ComponentModel;
using System.Data;
using System.Data.Common;
using TransitSoft.Model;
using TransitSoft.Persistance.dao;
using TransitSoft.Persistance.daoImpl.Util;

namespace TransitSoft.Persistance.daoImpl
{
    public class VehiculoDAOImpl : DAOImplBase, VehiculoDAO
    {
        private VehiculoDTO vehiculo;

        public VehiculoDAOImpl() : base("VEHICULO")
        {
            this.retornarLlavePrimaria = true;
            this.vehiculo = null;
        }

        protected override void ConfigurarListaDeColumnas()
        {
            this.listaColumnas.Add(new Columna("ID", true, true));
            this.listaColumnas.Add(new Columna("PLACA", false, false));
            this.listaColumnas.Add(new Columna("MARCA", false, false));
            this.listaColumnas.Add(new Columna("MODELO", false, false));
            this.listaColumnas.Add(new Columna("ANHO", false, false));
        }

        protected override void InstanciarObjetoDelResultSet(DbDataReader lector)
        {
            PropietarioDTO propietario = new PropietarioDTO();
            propietario.Id = lector.GetInt32(lector.GetOrdinal("ID_PROPIETARIO"));
            propietario.Dni = lector.GetString(lector.GetOrdinal("DNI"));
            propietario.Nombres = lector.GetString(lector.GetOrdinal("NOMBRES"));
            propietario.Apellidos = lector.GetString(lector.GetOrdinal("APELLIDOS"));
            propietario.Direccion = lector.GetString(lector.GetOrdinal("DIRECCION"));

            this.vehiculo = new VehiculoDTO();
            this.vehiculo.Id = lector.GetInt32(lector.GetOrdinal("ID_VEHICULO"));
            this.vehiculo.Placa = lector.GetString(lector.GetOrdinal("PLACA"));
            this.vehiculo.Marca = lector.GetString(lector.GetOrdinal("MARCA"));
            this.vehiculo.Modelo = lector.GetString(lector.GetOrdinal("MODELO"));
            this.vehiculo.Anho = lector.GetInt32(lector.GetOrdinal("ANHO"));
            this.vehiculo.Propietario = propietario;
        }

        protected override void LimpiarObjetoDelResultSet()
        {
            this.vehiculo = null;
        }

        protected override void AgregarObjetoALaLista(BindingList<Object> lista, DbDataReader lector)
        {
            this.InstanciarObjetoDelResultSet(lector);
            lista.Add(this.vehiculo);
        }

        public BindingList<VehiculoDTO> ListarPorPropietario(string descripcionPropietario)
        {
            string sql = this.GenerarSQLParaListarPorPropietario();
            BindingList<Object> lista;
            lista = base.ListarTodos(sql, this.IncluirValorDeParametrosParaListarPorPropietario, descripcionPropietario);
            BindingList<VehiculoDTO> retorno = new BindingList<VehiculoDTO>();
            foreach (VehiculoDTO objeto in lista)
            {
                retorno.Add(objeto);
            }
            return retorno;
        }

        private string GenerarSQLParaListarPorPropietario()
        {
            string sql = "SELECT V.ID AS ID_VEHICULO, ";
            sql += "V.PLACA, ";
            sql += "V.MARCA, ";
            sql += "V.MODELO, ";
            sql += "V.ANHO, ";
            sql += "P.ID AS ID_PROPIETARIO, ";
            sql += "P.DNI, ";
            sql += "P.NOMBRES, ";
            sql += "P.APELLIDOS, ";
            sql += "P.DIRECCION ";
            sql += "FROM vehiculo V ";
            sql += "JOIN vehiculo_propietario VP ON VP.ID_VEHICULO = V.ID ";
            sql += "JOIN propietario P ON P.ID = VP.ID_PROPIETARIO ";
            sql += "WHERE (P.DNI LIKE @p1 OR P.NOMBRES LIKE @p2 OR P.APELLIDOS LIKE @p3)";
            return sql;
        }

        private void IncluirValorDeParametrosParaListarPorPropietario(object objetoParametros)
        {
            string cadena = (string)objetoParametros;
            cadena = cadena.Trim();
            cadena = "%" + cadena + "%";
            AgregarParametro("@p1", cadena, this.comandoMySQL);
            AgregarParametro("@p2", cadena, this.comandoMySQL);
            AgregarParametro("@p3", cadena, this.comandoMySQL);

            AgregarParametro("@p1", cadena, this.comandoMSSQL);
            AgregarParametro("@p2", cadena, this.comandoMSSQL);
            AgregarParametro("@p3", cadena, this.comandoMSSQL);
        }
    }
}